internal/pkg/manifest/workload.go

```diff
@@ -441,14 +442,82 @@ func (p *PlacementArgs) isEmpty() bool {
 // PlacementString represents what types of subnets (public or private subnets) to place tasks.
 type PlacementString string
 
+// SecurityGroupsIDsOrConfig represents security groups attached to task. It supports unmarshalling
+// yaml which can either be of type SecurityGroupsConfig or a list of strings.
+type SecurityGroupsIDsOrConfig struct {
+	IDs            []string
+	AdvancedConfig SecurityGroupsConfig
+}
+
+func (s *SecurityGroupsIDsOrConfig) isEmpty() bool {
+	return len(s.IDs) == 0 && s.AdvancedConfig.isEmpty()
+}
+
+// SecurityGroupsConfig represents which security groups are attached to a task
+// and if default security group is applied.
+type SecurityGroupsConfig struct {
+	SecurityGroups []string `yaml:"groups"`
+	DenyDefault    *bool    `yaml:"deny_default"`
+}
+
+func (s *SecurityGroupsConfig) isEmpty() bool {
+	return len(s.SecurityGroups) == 0 && s.DenyDefault == nil
+}
+
+// UnmarshalYAML overrides the default YAML unmarshalling logic for the SecurityGroupsIDsOrConfig
+// struct, allowing it to perform more complex unmarshalling behavior.
+// This method implements the yaml.Unmarshaler (v3) interface.
+func (s *SecurityGroupsIDsOrConfig) UnmarshalYAML(value *yaml.Node) error {
+	if err := value.Decode(&s.AdvancedConfig); err != nil {
+		switch err.(type) {
+		case *yaml.TypeError:
+			break
+		default:
+			return err
+		}
+	}
+
+	if !s.AdvancedConfig.isEmpty() {
+		// Unmarshalled successfully to s.AdvancedConfig, unset s.IDs, and return.
+		s.IDs = nil
+		return nil
+	}
+
+	if err := value.Decode(&s.IDs); err != nil {
+		return errUnmarshalSecurityGroupOpts
+	}
+	return nil
+}
+
+// GetIDs returns security groups from SecurityGroupsIDsOrConfig that are attached to task.
+// nil is returned if no security groups are specified.
+func (s *SecurityGroupsIDsOrConfig) GetIDs() []string {
+	if !s.AdvancedConfig.isEmpty() && len(s.AdvancedConfig.SecurityGroups) > 0 {
+		return s.AdvancedConfig.SecurityGroups
+	}
+	if len(s.IDs) > 0 {
+		return s.IDs
+	}
+	return nil
+}
+
+// IsDefaultSecurityGroupDenied returns true if DenyDefault is set to true
+// in SecurityGroupsIDsOrConfig.AdvancedConfig. Otherwise, false is returned.
+func (s *SecurityGroupsIDsOrConfig) IsDefaultSecurityGroupDenied() bool {
+	if !s.AdvancedConfig.isEmpty() && s.AdvancedConfig.DenyDefault != nil {
+		return *s.AdvancedConfig.DenyDefault
+	}
+	return false
```

Comment:
nit: this should be effectively the same as yours but is more aligned with the rest of the codebase! `aws.BoolValue` returns the referenced value given a pointer, and `false` if the pointer is `nil`

```suggestion
	if !s.AdvancedConfig.isEmpty() {
		return aws.BoolValue(s.AdvancedConfig.DenyDefault)
	}
	return false
```